### PR TITLE
Store file contents as raw vectors for binary files

### PR DESCRIPTION
**Pull request by gaborcsardi**
We define binary files as files containing at least one embedded null
character.

All files are read as raw vectors now. Binary files are kept as raw
vectors, text files as character. As a typical plugin either works on
binary files, or on text files, but not both, this is fine, we
don't need to convert between the raw and text representation.

### CODE REVIEW

File: src/read_file.cpp [hunk 1]
```diff
@@ -7,13 +7,15 @@ using namespace Rcpp;
...
 //' @useDynLib rsmith
 // [[Rcpp::export]]
-CharacterVector read_file(std::string path) {
-  std::ifstream in(path.c_str());
-  std::string contents;
-  in.seekg(0, std::ios::end);
-  contents.resize(in.tellg());
+RawVector read_file_raw(std::string path) {
+  std::ifstream in(path.c_str(), std::ios::binary | std::ios::ate);
+  size_t file_size=in.tellg();
```
Comment (hadley) [inline src/read_file.cpp]: Could you put spaces around = please?

Comment (gaborcsardi) [inline src/read_file.cpp]: Sure, done.

File: R/output.R [hunk 2]
```diff
@@ -35,7 +35,11 @@ preview <- function(rsmith) {
 
   for(file in files) {
     message(file$metadata$.path)
-    cat(file$content, "\n", sep = "")
+    if (is_binary(file)) {
+      cat("Binary file not shown\n")
```
Comment (hadley) [inline R/output.R]: I'd make this "<binary file>"

File: R/rsmith-file.R [hunk 3]
```diff
@@ -1,3 +1,12 @@
+read_file <- function(path) {
+  contents <- read_file_raw(path)
+  if (is_binary_raw(contents)) {
+    contents
+  } else {
+    rawToChar(contents)
```
Comment (hadley) [inline R/rsmith-file.R]: I suspect at some point we'll need to worry about encoding issues here.  Or maybe we should just make the Rmarkdown policy that all text files at utf-8?

File: R/output.R (hunk 2 above)
Comment (gaborcsardi) [inline R/output.R]: You mean, to say nothing if it is a binary file?

File: R/rsmith-file.R (hunk 3 above)
Comment (gaborcsardi) [inline R/rsmith-file.R]: I like making utf-8 a requirement. Much simpler for everyone, and people should use utf-8 for new stuff, anyway.

File: R/output.R (hunk 2 above)
Comment (hadley) [inline R/output.R]: Oops, missed some backticks, I meant: `"<binary file>"`

Comment (gaborcsardi) [inline R/output.R]: Yes, that makes more sense. :) Done.

File: R/rsmith-file.R (hunk 3 above)
Comment (hadley) [inline R/rsmith-file.R]: Then I think we need

``` r
text <- rawToChar(contents)
Encoding(text) <- "UTF-8"
text
```

Comment (gaborcsardi) [inline R/rsmith-file.R]: Done.

Comment (gaborcsardi) [inline R/rsmith-file.R]: On a second thought, if we mean this to be a general data/file processing tool, then it is desirable to support other encodings as well. But I think we can add a way to specify file encodings later. For now UTF-8 is fine, that would be the default if the encoding is not specified, anyway.

**Comment by gaborcsardi** [comment]
I'll start rewriting rsmith using R6 soon. Any chance that you can merge this PR?

**Comment by hadley** [comment]
Done - thanks!

**Pull request merged**